### PR TITLE
Fix dynamic insertion of wrapped or redistributing content.

### LINKED ISSUE #1631: [0.9.4] Dom-if template doesn't stamp when its content contains a wrapped insertion point
The following case demonstrates the issue:
http://jsbin.com/lolareyeca/1/edit?html,output

Clicking *Go* should cause the dom-if template to stamp, but it doesn't.
When removing the **div** element around the insertion point, it works as intended.

Probably related to #1565, #1584?

### LINKED ISSUE #1753: Content nodes in `dom-if` template do not distribute correctly
The following HTML page causes a TypeError when loaded:

```html
<!DOCTYPE html>
<html>
  <head>
    <script>
var Polymer = Polymer || {};
// Uncomment and the error disappears.
// Polymer.dom = 'shadow';
    </script>
    <script src="../webcomponentsjs/webcomponents-lite.js"></script>
    <link rel="import" href="../polymer/polymer.html">
  </head>
  <body>

    <!-- Define test element -->
    <dom-module id="test-elem1">
      <template>
        <content></content>
      </template>
    </dom-module>

    <dom-module id="test-elem2">
      <template>
        <test-elem1>
          <template is="dom-if" if="true">
            <content></content>
          </template>
        </test-elem1>
      </template>
    </dom-module>
    <script>
Polymer({
  is: 'test-elem1',
});

Polymer({
  is: 'test-elem2',
});
    </script>
    <!--
      if using native shadow DOM, works correctly, showing "Hello, World!".
      with shady DOM, causes a TypeError in clearDistributedDestinationInsertionPoints().
    -->
    <test-elem2 id="foo">
      <div>Hello, World!</div>
    </test-elem2>
  </body>
</html>
```

I've tested this in Chrome. Firefox was a little pickier with my code, so I can't say for certain that it also raises an error, but I suspect so.

### LINKED ISSUE #1800: <template is=dom-if> fails to add rows to a table if they contain <content>
Demo at http://plnkr.co/edit/adwhKJcfSxiM3tjxU6JS?p=preview. 

```html
    <table>
      <template is="dom-if" if="{{hasSpan}}">
        <tr>
          <th>Header</th><td><content select="span"></content></td>
        </tr>
      </template>
    ...
```
never shows the row, but replacing the `<content>` with anything else makes it work.

### PR DESCRIPTION
**Pull request by kevinpschaaf**
Fixes #1631.
Fixes #1753.
Fixes #1800.

### CODE REVIEW

File: src/lib/dom-api.html [hunk 1]
```diff
@@ -173,17 +166,39 @@
...
+        // There are 2 possible cases where a host may need distribution:
+        // 1. <content> being inserted (the host of the shady root where
+        //    content is inserted needs distribution)
+        // 2. children being inserted into parent with a shady root (parent
+        //    needs distribution)
+        // Note that when 1 & 2 are both true (re-distribution), distributing the
+        // <content>'s host (1) will cause the parent to distribute as well (2)
+        var host;
+        if (isContent || hasContent) {
+          host = this._ownerShadyRootForNode(parent).host;
```
Comment (sorvell) [inline src/lib/dom-api.html]: it's possible we don't have an owner root here isn't it?

File: src/lib/dom-api.html [hunk 2]
```diff
@@ -173,17 +166,39 @@
...
+        if (host) {
           this._lazyDistribute(host);
         }
-        return distribute;
+        // Return true when distribution will fully handle the composition
+        // Note that if a content was being inserted that was wrapped by a node,
+        // and the parent does not need distribution, return false to allow
+        // the nodes to be added directly, after which children may be
+        // distributed and composed into the wrapping node(s)
+        var wrappedContent = hasContent &&
```
Comment (sorvell) [inline src/lib/dom-api.html]: hasContent seems like a Boolean, let's make the name more explicit

File: src/lib/dom-api.html [hunk 3]
```diff
@@ -173,17 +166,39 @@
...
+        //    content is inserted needs distribution)
+        // 2. children being inserted into parent with a shady root (parent
+        //    needs distribution)
+        // Note that when 1 & 2 are both true (re-distribution), distributing the
+        // <content>'s host (1) will cause the parent to distribute as well (2)
+        var host;
+        if (isContent || hasContent) {
+          host = this._ownerShadyRootForNode(parent).host;
           this._updateInsertionPoints(host);
+        } else if (this._parentNeedsDistribution(parent)) {
```
Comment (sorvell) [inline src/lib/dom-api.html]: this check is always needed, remove the `else`.

**Comment by sorvell** [comment]
LGTM

**Comment by pflannery** [comment]
I found that when using a dom-bind template its still failing (dom-if present or not). http://jsbin.com/vifodowilu/edit?html,output

```html
  <template id="main" is="dom-bind">
      <test-elem2 id="foo">
        <div>Hello, World!</div>
      </test-elem2>
  </template>
```

**Pull request merged**